File: sevens/sevens.py
from random import shuffle, seed


# card object
class Card:
    def __init__(self, suit, rank, play=0):
        self.suit = suit
        self.rank = rank
        self.connect = -1
        self.shield = -1

    def __str__(self):
        return "{} of {}".format(self.rank, self.suit)

    def __repr__(self):
        return "{} of {}".format(self.rank, self.suit)

    def __eq__(self, other):
        return self.__dict__ == other.__dict__


def createDeck():
    deck = []
    suit = ["clubs", "spades", "diamonds", "hearts"]
    rank = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]
    for i in suit:
        for j in rank:
            if j == 7:
                deck.append(Card(i, j))
            elif j > 7:
                deck.append(Card(i, j))
            elif j < 7:
                deck.append(Card(i, j))

    return deck


def playableCards(hand, played):
    playable = []
    for i in hand:
        if i.rank == 7:
            playable.append(i)
            continue
        for j in played:
            if i.suit == j.suit:
                if i.rank > 7 and j.rank == i.rank - 1:
                    playable.append(i)
                elif i.rank < 7 and j.rank == i.rank + 1:
                    playable.append(i)

    return playable


def connected(hand):
    # -1: not evaluated
    # 0: not connected
    # 1: singlely connected
    # 3: tail
    for i in range(len(hand) - 1):
        # Check tail end
        if hand[i].rank == 1 or hand[i].rank == 13:
            hand[i].connect = 3
        elif hand[i + 1].rank == hand[i].rank + 1:
            hand[i].connect = 1
            hand[i + 1].connect = 1
        elif hand[i].connect != 1:
            hand[i].connect = 0
    # Last card
    if hand[-1].connect == -1:
        if hand[-1].rank == 1 or hand[-1].rank == 13:
            hand[-1].connect = 3
        elif hand[-1].connect != 1:
            hand[-1].connect = 0


def shield(hand):
    clubs = [x for x in hand if x.suit == "clubs"]
    spades = [x for x in hand if x.suit == "spades"]
    diamonds = [x for x in hand if x.suit == "diamonds"]
    hearts = [x for x in hand if x.suit == "hearts"]
    if clubs[0].rank != 1:
        clubs[0].shield = 1
    if clubs[-1].rank != 13:
        clubs[-1].shield = 1
    if spades[0].rank != 1:
        spades[0].shield = 1
    if spades[-1].rank != 13:
        spades[-1].shield = 1
    if diamonds[0].rank != 1:
        diamonds[0].shield = 1
    if diamonds[-1].rank != 13:
        diamonds[-1].shield = 1
    if hearts[0].rank != 1:
        hearts[0].shield = 1
    if hearts[-1].rank != 13:
        hearts[-1].shield = 1


# Minimax heuristic which values connected cards
def control(p1, p2, pile, first, depth):
    # Heuristic
    p1_plays = playableCards(p1, pile)
    p2_plays = playableCards(p2, pile)
    if depth == 0:
        eval = 0
        for i in p1:
            if i.connect == 1:
                eval += 3
            elif i.connect == 3:
                eval -= 10
        for i in p2:
            if i.connect == 1:
                eval -= 3
            elif i.connect == 3:
                eval += 10
        return eval
    if not p1:
        return 10000
    elif not p2:
        return -10000

    if first:
        score = -10000
        if not p1_plays:
            return control(p1, p2, pile, False, depth - 1)
        for i in p1_plays:
            new_pile = pile.copy()
            new_p1 = p1.copy()
            new_pile.append(i)
            new_p1.remove(i)
            score = max(score, control(new_p1, p2, new_pile, False, depth - 1))
    else:
        score = 10000
        if not p2_plays:
            return control(p1, p2, pile, True, depth - 1)
        for i in p2_plays:
            new_pile = pile.copy()
            new_p2 = p2.copy()
            new_pile.append(i)
            new_p2.remove(i)
            score = min(score, control(p1, new_p2, new_pile, True, depth - 1))

    return score


# Minimax alpha-beta heuristic which values connected cards
def controlab(p1, p2, pile, first, depth, alpha, beta):
    # Heuristic
    p1_plays = playableCards(p1, pile)
    p2_plays = playableCards(p2, pile)
    if depth == 0:
        eval = 0
        for i in p1:
            if i.connect == 1:
                eval += 3
            elif i.connect == 3:
                eval -= 10
        for i in p2:
            if i.connect == 1:
                eval -= 3
            elif i.connect == 3:
                eval += 10
        return eval
    if not p1:
        return 10000
    elif not p2:
        return -10000

    if first:
        score = -10000
        if not p1_plays:
            return controlab(p1, p2, pile, False, depth - 1, alpha, beta)
        for i in p1_plays:
            new_pile = pile.copy()
            new_p1 = p1.copy()
            new_pile.append(i)
            new_p1.remove(i)
            score = max(score, controlab(new_p1, p2, new_pile, False, depth - 1, alpha, beta))
            alpha = max(alpha, score)
            if alpha >= beta:
                break
    else:
        score = 10000
        if not p2_plays:
            return controlab(p1, p2, pile, True, depth - 1, alpha, beta)
        for i in p2_plays:
            new_pile = pile.copy()
            new_p2 = p2.copy()
            new_pile.append(i)
            new_p2.remove(i)
            score = min(score, controlab(p1, new_p2, new_pile, True, depth - 1, alpha, beta))
            beta = min(beta, score)
            if alpha >= beta:
                break

    return score


def controlabSup(p1, p2, pile, first, depth, alpha, beta):
    # Heuristic
    p1_plays = playableCards(p1, pile)
    p2_plays = playableCards(p2, pile)
    if depth == 0:
        eval = 0
        for i in p1:
            if i.connect == 1:
                eval += 2
            if i.connect == 3:
                eval -= 15
            if i.shield == 1:
                eval += 5
        for i in p2:
            if i.connect == 1:
                eval -= 2
            if i.connect == 3:
                eval += 15
            if i.shield == 1:
                eval -= 5
        return eval
    if not p1:
        return 10000
    elif not p2:
        return -10000

    if first:
        score = -10000
        if not p1_plays:
            return controlabSup(p1, p2, pile, False, depth - 1, alpha, beta)
        for i in p1_plays:
            new_pile = pile.copy()
            new_p1 = p1.copy()
            new_pile.append(i)
            new_p1.remove(i)
            score = max(score, controlabSup(new_p1, p2, new_pile, False, depth - 1, alpha, beta))
            alpha = max(alpha, score)
            if alpha >= beta:
                break
    else:
        score = 10000
        if not p2_plays:
            return controlabSup(p1, p2, pile, True, depth - 1, alpha, beta)
        for i in p2_plays:
            new_pile = pile.copy()
            new_p2 = p2.copy()
            new_pile.append(i)
            new_p2.remove(i)
            score = min(score, controlabSup(p1, new_p2, new_pile, True, depth - 1, alpha, beta))
            beta = min(beta, score)
            if alpha >= beta:
                break

    return score


# Minimax heuristic which values number of possible plays
def plays(p1, p2, pile, first, depth):
    p1_plays = playableCards(p1, pile)
    p2_plays = playableCards(p2, pile)
    # Heuristic
    if depth == 0:
        return len(p1_plays) - len(p2_plays)
    if not p1:
        return 10000
    elif not p2:
        return -10000

    if first:
        score = -10000
        if not p1_plays:
            return plays(p1, p2, pile, False, depth - 1)
        for i in p1_plays:
            new_pile = pile.copy()
            new_p1 = p1.copy()
            new_pile.append(i)
            new_p1.remove(i)
            score = plays(new_p1, p2, new_pile, False, depth - 1)
    else:
        score = 10000
        if not p2_plays:
            return plays(p1, p2, pile, True, depth - 1)
        for i in p2_plays:
            new_pile = pile.copy()
            new_p2 = p2.copy()
            new_pile.append(i)
            new_p2.remove(i)
            score = plays(p1, new_p2, new_pile, True, depth - 1)

    return score


def evaluate(p1, p2, pile, first, depth):
    if first:
        p1_plays = playableCards(p1, pile)
        print("P1: " + str(p1_plays))
        score = -20000
        best_move = 0
        for i in p1_plays:
            new_pile = pile.copy()
            new_p1 = p1.copy()
            new_pile.append(i)
            new_p1.remove(i)
            cur_score = max(score, controlabSup(new_p1, p2, new_pile, False, 11, -10000, 10000))
            if cur_score > score:
                score = cur_score
                best_move = i
            print(str(i) + ": " + str(cur_score))

        print(str(best_move) + ": index " + str(p1.index(best_move)))
    else:
        p2_plays = playableCards(p2, pile)
        print("P2: " + str(p2_plays))
        score = 20000
        best_move = 0
        for i in p2_plays:
            new_pile = pile.copy()
            new_p2 = p2.copy()
            new_pile.append(i)
            new_p2.remove(i)
            cur_score = min(score, controlabSup(p1, new_p2, new_pile, True, 11, -10000, 10000))
            if cur_score < score:
                score = cur_score
                best_move = i
            print(str(i) + ": " + str(cur_score))

        print(str(best_move) + ": index " + str(p2.index(best_move)))
    return best_move


def playCard(p1, p2, card, pile):
    if card in p1:
        pile.append(card)
        p1.remove(card)
    elif card in p2:
        pile.append(card)
        p2.remove(card)


'''
def player():
    print("Play a card")
        p1_plays = playableCards(p1, pile)
        if not playableCards(p1, pile):
            continue
        print(p1_plays)
        i = int(input())
        while i not in list(range(len(p1_plays))):
            print("Try again")
            i = int(input())
        playCard(p1, p2, p1_plays[i], pile)
'''


def main():
    # Seed 1: P1 gets 6 tails
    # Seed 30: P1 gets 4 7s
    # Seed 42: P1 gets 5 tails, P2 gets 3 7s (P2 forced win?)
    seed(42)
    deck = createDeck()
    shuffle(deck)
    p1 = deck[0:26]
    p2 = deck[26:52]
    pile = []
    p1.sort(key=lambda x: (x.suit, x.rank))
    p2.sort(key=lambda x: (x.suit, x.rank))
    seven = Card("diamonds", 7)
    first = True
    for i in p1:
        if i == seven:
            seven = i
            first = False
            break
    playCard(p1, p2, seven, pile)
    connected(p1)
    connected(p2)
    shield(p1)
    shield(p2)

    while p1 or p2:
        if first:
            if not playableCards(p1, pile):
                first = False
                continue
            move = evaluate(p1, p2, pile, True, 11)
            playCard(p1, p2, move, pile)
            first = False
        else:
            if not playableCards(p2, pile):
                first = True
                continue
            move = evaluate(p1, p2, pile, False, 11)
            playCard(p1, p2, move, pile)
            first = True


main()

File: sevens/sevens.js
// Select canvas by HTML tag
var canvas = document.querySelector("canvas");
canvas.width = 1800;
canvas.height = 1200;
canvas.style.width = "900px";
canvas.style.height = "600px";

var ctx = canvas.getContext('2d')
ctx.scale(2, 2);
ctx.imageSmoothingEnabled = false;

card_width = 500;
card_height = 726;

// Variable for storing mouse coordinates
var mouse = {
    x: undefined,
    y: undefined
};

// Variable for storing mouse click coordinates
var mousedown = {
    x: undefined,
    y: undefined
};

window.addEventListener("mousedown", function(m) {
    mousedown.x = m.x;
    mousedown.y = m.y;
});

window.addEventListener("mousemove", function(m) {
    mouse.x = m.x;
    mouse.y = m.y;
});

window.addEventListener("mouseup", function() {
    mousedown.x = undefined;
    mousedown.y = undefined;
    for (var i = 0; i < p1_plays.length; i++) {
        p1_plays[i].xcorner = p1_plays[i].x + card_width/8;
        p1_plays[i].ycorner = p1_plays[i].y + card_height/8;
        if (p1_plays[i].xTarget > p1_plays[i].xcorner ||
            p1_plays[i].xTarget + card_width/8 < p1_plays[i].x ||
            p1_plays[i].yTarget > p1_plays[i].ycorner ||
            p1_plays[i].yTarget + card_height/8 < p1_plays[i].y && p1_plays[i].dragging == false) {
                p1_plays[i].x = p1_plays[i].xOld;
                p1_plays[i].y = p1_plays[i].yOld;
            } else {
                p1_plays[i].x = p1_plays[i].xTarget;
                p1_plays[i].y = p1_plays[i].yTarget;
                pile.push(p1_plays[i]);
                p1.splice(p1.indexOf(p1_plays[i]), 1);
                p1_plays.splice(i, 1);
                p1Move = false;
                return;
            }
    }
});

// Shuffle Algorithm
function shuffle(a) {
    var j, x, i;
    for (i = a.length - 1; i > 0; i--) {
        j = Math.floor(Math.random() * (i + 1));
        x = a[i];
        a[i] = a[j];
        a[j] = x;
    }
    return a;
}

// Suit sorting
function compareSuit(a, b) {
    if (a.suit < b.suit)
      return -1;
    if (a.suit > b.suit)
      return 1;
    return 0;
}

// Rank sorting
function compareRank(a, b) {
    if (a.rank < b.rank)
      return -1;
    if (a.rank > b.rank)
      return 1;
    return 0;
}

// Rank sorting (desending)
function compareRankD(a, b) {
    if (a.rank > b.rank)
      return -1;
    if (a.rank < b.rank)
      return 1;
    return 0;
}

function searchRankSuit(hand, rank, suit) {
    for (var i = 0; i < hand.length; i++) {
        if (hand[i].rank == rank && hand[i].suit == suit)
            return true;
    }
    return false;
}

// Card object
function Card(suit, rank, src) {
    this.suit = suit;
    this.rank = rank;
    this.sprite = src;
    // x, y positions
    this.x;
    this.y;
    // Corner positions
    this.xcorner;
    this.ycorner;
    this.dragging = false;
    // Original x, y corners when dragging
    this.dragx = undefined;
    this.dragy = undefined;
    // x, y position for final position of card
    this.xTarget;
    this.yTarget;
    // x, y position before dragging
    this.xOld;
    this.yOld;
    this.connect = -1;

    this.draw = function() {
        this.xcorner = this.x + card_width/8;
        this.ycorner = this.y + card_height/8;
        var img = new Image();
        img.src = this.sprite;
        ctx.drawImage(img, this.x, this.y, img.width/8, img.height/8);
    }
}

// Create deck
function generateDeck() {
    var deck = [];
    for (var i = 0; i < 4; i++) {
        for (var j = 1; j <= 13; j++) {
            var suit;
            switch(i) {
                case 0:
                    suit = "club";
                    break;
                case 1:
                    suit = "diamond";
                    break;
                case 2:
                    suit = "heart";
                    break;
                case 3:
                    suit = "spade";
                    break;
            }
            var card = new Card(suit, j, "Cards/" + (13 * i + j) + ".png");
            if (card.rank == 7) {
                card.xTarget = 310 + i * 70;
                card.yTarget = 250;
            } else if (card.rank < 7) {
                card.xTarget = 310 + i * 70;
                card.yTarget = 20 + 20 * j;
            } else {
                card.xTarget = 310 + i * 70;
                card.yTarget = 360 + 20 * (j - 8);
            }
            deck.push(card);
        }
    }
    return deck
}

// Set position of cards for CPU
function setCPU(hand) {
    var suits = ["club", "diamond", "heart", "spade"];
    suits.reverse();
    var lower = [];
    var upper = [];
    for (var i = 0; i < hand.length; i++) {
        if (hand[i].rank <= 7) {
            lower.push(hand[i]);
        } else {
            upper.push(hand[i]);
        }
    }
    lower.sort(compareRank);
    lower.sort(compareSuit);
    upper.sort(compareRankD);
    upper.sort(compareSuit);
    var drawn;
    for (var i = 0; i < 4; i++) {
        drawn = 0;
        for (var j = 0; j < lower.length; j++) {
            if (lower[j].suit == suits[i]) {
                lower[j].x = 835 - i * 70
                lower[j].y = 80 + 20 * drawn
                drawn++;
            }
        }
    }
    upper.reverse()
    for (var i = 0; i < 4; i++) {
        drawn = 0;
        for (var j = upper.length - 1; j >= 0; j--) {
            if (upper[j].suit == suits[i]) {
                upper[j].x = 835 - i * 70
                upper[j].y = 360 + 20 * drawn
                drawn++;
            }
        }
    }
    var arr = [lower, upper];
    return arr;
}

// Set X, Y positions of cards for player
function setPlayer(hand) {
    var suits = ["club", "diamond", "heart", "spade"];
    var lower = [];
    var upper = [];
    // Partition hand
    for (var i = 0; i < hand.length; i++) {
        if (hand[i].rank <= 7) {
            lower.push(hand[i]);
        } else {
            upper.push(hand[i]);
        }
    }
    // Sort by rank and suit
    lower.sort(compareRank);
    lower.sort(compareSuit);
    upper.sort(compareRankD);
    upper.sort(compareSuit);
    var drawn;
    // Set positions
    for (var i = 0; i < 4; i++) {
        drawn = 0;
        for (var j = 0; j < lower.length; j++) {
            if (lower[j].suit == suits[i]) {
                lower[j].x = i * 70
                lower[j].y = 80 + 20 * drawn
                lower[j].xOld = i * 70
                lower[j].yOld = 80 + 20 * drawn
                drawn++;
            }
        }
    }
    upper.reverse()
    for (var i = 0; i < 4; i++) {
        drawn = 0;
        for (var j = upper.length - 1; j >= 0; j--) {
            if (upper[j].suit == suits[i]) {
                upper[j].x = i * 70
                upper[j].y = 360 + 20 * drawn
                upper[j].xOld = i * 70
                upper[j].yOld = 360 + 20 * drawn
                drawn++;
            }
        }
    }
    var arr = [lower, upper];
    return arr;
}

function drawCPU(hand) {
    for (var i = 0; i < hand[0].length; i++) {
        hand[0][i].draw();
    }
    for (var i = hand[1].length - 1; i >= 0; i--) {
        hand[1][i].draw();
    }
}

function drawPlayer(hand, plays) {
    for (var i = 0; i < hand[0].length; i++) {
        hand[0][i].draw();
    }
    for (var i = hand[1].length - 1; i >= 0; i--) {
        hand[1][i].draw();
    }
    for (var i = 0; i < plays.length; i++) {
        plays[i].draw();
    }
}

function drawGrid() {
    for (var i = 0; i < 4; i++) {
        ctx.beginPath();
        ctx.strokeStyle = "black";
        ctx.lineWidth = "1";
        // Upper cards
        ctx.rect(310 + i * 70, 140, card_width/8, card_height/8);
        // Center cards
        ctx.rect(310 + i * 70, 250, card_width/8, card_height/8);
        // Lower cards
        ctx.rect(310 + i * 70, 360, card_width/8, card_height/8);
        ctx.stroke();
        ctx.closePath();
    }
}

function dragCard(mousedown, mouse) {
    // Check if mouse is held down
    if (mousedown.x == undefined) {
        for (var i = 0; i < p1_plays.length; i++) {
            p1_plays[i].dragging = false;
        }
        return;
    }
    // Move card if mouse is down and over a card
    for (var i = 0; i < p1_plays.length; i++) {
        if (p1_plays[i].dragging == true) {
            p1_plays[i].x = p1_plays[i].dragx - card_width/8 + (mouse.x - mousedown.x);
            p1_plays[i].y = p1_plays[i].dragy - card_height/8 + (mouse.y - mousedown.y);
            return;
        }
    }
    // Check if over card on mouse click
    for (var i = 0; i < p1_plays.length; i++) {
        if (mousedown.x > p1_plays[i].x && mousedown.x < p1_plays[i].xcorner && mousedown.y > p1_plays[i].y && mousedown.y < p1_plays[i].ycorner) {
            p1_plays[i].dragging = true;
            p1_plays[i].xOld = p1_plays[i].x;
            p1_plays[i].yOld = p1_plays[i].y;
            p1_plays[i].dragx = p1_plays[i].xcorner;
            p1_plays[i].dragy = p1_plays[i].ycorner;
            return;
        }
    }
}

function playableCards(hand, pile) {
    var playable = []
    for (var i = 0; i < hand.length; i++) {
        if (hand[i].rank == 7) {
            playable.push(hand[i]);
            continue;
        }   
        for (var j = 0; j < pile.length; j++) {
            if (hand[i].suit == pile[j].suit) {
                if (hand[i].rank > 7 && pile[j].rank == hand[i].rank - 1) {
                    playable.push(hand[i]);
                } else if (hand[i].rank < 7 && pile[j].rank == hand[i].rank + 1) {
                    playable.push(hand[i]);
                }
            }
        }
    }
    return playable
}

function connected(hand) {
    // -1: not evaluated
    // 0: not connected
    // 1: singlely connected
    // 3: tail
    for (var i = 0; i < hand.length - 1; i++) {
        // Check tail end
        if (hand[i].rank == 1 || hand[i].rank == 13) {
            hand[i].connect = 3;
        } else if (hand[i + 1].rank == hand[i].rank + 1) {
            hand[i].connect = 1;
            hand[i + 1].connect = 1;
        } else if (hand[i].connect != 1) {
            hand[i].connect = 0;
        }
    }
    if (hand[hand.length - 1].rank == 1 || hand[hand.length - 1].rank == 13) {
        hand[hand.length - 1].connect = 3;
    } else if (hand[hand.length - 1].connect != 1) {
        hand[hand.length - 1].connect = 0;
    }
}

function minimax(p1, p2, pile, first, depth, alpha, beta) {
    // Heuristic
    var p1_plays = playableCards(p1, pile);
    var p2_plays = playableCards(p2, pile);
    // Tail-end Heuristic
    /*
    if (depth == 0) {
        var eval = 0;
        for (var i = 0; i < p1.length; i++) {
            if (p1[i].connect == 1)
                eval += 1;
            else if (p1[i].connect == 3 && !p1_plays.includes(p1[i])) 
                eval -= 5;
        }
        for (var i = 0; i < p2.length; i++) {
            if (p2[i].connect == 1)
                eval -= 1;
            else if (p2[i].connect == 3 && !p2_plays.includes(p1[i]))
                eval += 5;
        }
        return eval;
    }
    */
    // Control heuristic
    if (depth == 0) {
        var eval = 0;
        for (var i = 0; i < p1_plays.length; i++) {
            if (p1_plays[i].connect == 3)
                eval++;
            else if (p1_plays[i].rank < 7) {
                var rank = p1_plays[i].rank - 1;
                while (rank > 0) {
                    if (searchRankSuit(p1, rank, p1_plays[i].suit)) {
                        rank--;
                        eval++;
                    } else {
                        break;
                    }
                }
            } else if (p1_plays[i].rank > 7) {
                var rank = p1_plays[i].rank + 1;
                while (rank < 14) {
                    if (searchRankSuit(p1, rank, p1_plays[i].suit)) {
                        rank++;
                        eval++;
                    } else {
                        break;
                    }
                }
            } else if (p1_plays[i].rank == 7) {
                if (searchRankSuit(p1, 6, p1_plays[i].suit) && searchRankSuit(p1, 8, p1_plays[i].suit)) {
                    eval++;
                } else {
                    break;
                }
            }
        }
        for (var i = 0; i < p2_plays.length; i++) {
            if (p2_plays[i].connect == 3)
                eval--;
            else if (p2_plays[i].rank < 7) {
                var rank = p2_plays[i].rank - 1;
                while (rank > 0) {
                    if (searchRankSuit(p2, rank, p2_plays[i].suit)) {
                        rank--;
                        eval--;
                    } else {
                        break;
                    }
                }
            } else if (p2_plays[i].rank > 7) {
                var rank = p2_plays[i].rank + 1;
                while (rank < 14) {
                    if (searchRankSuit(p2, rank, p2_plays[i].suit)) {
                        rank++;
                        eval--;
                    } else {
                        break;
                    }
                }
            } else if (p2_plays[i].rank == 7) {
                if (searchRankSuit(p2, 6, p2_plays[i].suit) && searchRankSuit(p2, 8, p2_plays[i].suit)) {
                    eval--;
                } else {
                    break;
                }
            }
        }
        return eval;
    }
    if (p1 === undefined || p1.length == 0)
        return 10000;
    else if (p2 === undefined || p2.length == 0)
        return -10000;
    
    if (first == true) {
        var score = -10000;
        if (p1_plays === undefined || p1_plays.length == 0)
            return minimax(p1, p2, pile, false, depth - 1, alpha, beta);
        for (var i = 0; i < p1_plays.length; i++) {
            var newPile = pile.slice();
            var newP1 = p1.slice();
            newPile.push(p1_plays[i]);
            newP1.splice(newP1.indexOf(p1_plays[i]), 1);
            score = Math.max(score, minimax(newP1, p2, newPile, false, depth - 1, alpha, beta));
            alpha = Math.max(alpha, score);
            if (alpha >= beta)
                break;
        }
    } else {
        var score = 10000;
        if (p2_plays === undefined || p2_plays.length == 0)
            return minimax(p1, p2, pile, true, depth - 1, alpha, beta);
        for (var i = 0; i < p2_plays.length; i++) {
            var newPile = pile.slice();
            var newP2 = p2.slice();
            newPile.push(p2_plays[i]);
            newP2.splice(newP2.indexOf(p2_plays[i]), 1);
            score = Math.min(score, minimax(p1, newP2, newPile, true, depth - 1, alpha, beta));
            beta = Math.min(beta, score);
            if (alpha >= beta)
                break;
        }
    }
    return score;
}

function evaluate(p1, cpu, first, pile, depth) {
    var p1Plays = playableCards(p1, pile);
    var cpuPlays = playableCards(cpu, pile);
    if (cpuPlays === undefined || cpuPlays.length == 0)
        return
    var bestMove;
    if (first) {
        var score = -20000;
        for (var i = 0; i < p1Plays.length; i++) {
            var newPile = pile.slice();
            var newP1 = p1.slice();
            newPile.push(p1Plays[i]);
            newP1.splice(newP1.indexOf(p1Plays[i]), 1);
            var cur_score = Math.max(score, minimax(newP1, cpu, newPile, false, depth, -10000, 10000));
            if (cur_score > score) {
                score = cur_score;
                bestMove = cpuPlays[i];
            } 
        }
        return score;
    } else {
        for (var i = 0; i < cpuPlays.length; i++) {
            var score = 20000;
            var newPile = pile.slice();
            var newCPU = cpu.slice();
            newPile.push(cpuPlays[i]);
            newCPU.splice(newCPU.indexOf(cpuPlays[i]), 1);
            var cur_score = Math.min(score, minimax(p1, newCPU, newPile, true, depth, -10000, 10000));
            if (cur_score < score) {
                score = cur_score;
                bestMove = cpuPlays[i];
            } 
        }
        pile.push(bestMove);
        cpu.splice(cpu.indexOf(bestMove), 1);
        // Change position
        bestMove.x = bestMove.xTarget;
        bestMove.y = bestMove.yTarget;
        return score;
    }
}

function blueBorder(plays) {
    for (var i = 0; i < plays.length; i++) {
        ctx.beginPath();
        ctx.strokeStyle = "blue";
        ctx.lineWidth = "2";
        ctx.rect(plays[i].x, plays[i].y, card_width/8, card_height/8);
        ctx.stroke()
        ctx.closePath();
    }
}

var deck = generateDeck();
shuffle(deck);
// Boolean to check if player moves
var p1Move;
var p1 = deck.slice(0, 26);
var p2 = deck.slice(26, 52);
connected(p1);
connected(p2);
p1.sort(compareRank);
p2.sort(compareRank);
var pile = [];
var p1_draw = setPlayer(p1);
var p2_draw = setCPU(p2);
// Search and play 7 of diamonds
for (var i = 0; i < p1.length; i++) {
    if (p1[i].suit == "diamond" && p1[i].rank == 7) {
        pile.push(p1[i]);
        p1[i].x = p1[i].xTarget;
        p1[i].y = p1[i].yTarget;
        p1.splice(i, 1);
        p1Move = false;
        break;
    }
}
if (pile.length != 1) {
    for (var i = 0; i < p2.length; i++) {
        if (p2[i].suit == "diamond" && p2[i].rank == 7) {
            pile.push(p2[i]);
            p2[i].x = p2[i].xTarget;
            p2[i].y = p2[i].yTarget;
            p2.splice(i, 1);
            p1Move = true;
            break;
        }
    }
}
var p1_plays = playableCards(p1, pile);
if (p1_plays === undefined || p1_plays.length == 0) {
    p1Move = false
}

var toggle = true;
var score = evaluate(p1, p2, true, pile, 11);
function update() {
    // Set to 30fps
    requestAnimationFrame(update);
    if (toggle) {
        ctx.clearRect(0, 0, canvas.width, canvas.height);
        ctx.fillStyle = "green";
        ctx.fillRect(0, 0, innerWidth, innerHeight);
        drawGrid();
        while (p1Move == false) {
            score = evaluate(p1, p2, false, pile, 13);
            p1_plays = playableCards(p1, pile);
            if (p1_plays === undefined || p1_plays.length == 0) {
                if (pile.length == 52) {
                    p1Move = true;
                }
            } else {
                p1Move = true
            }  
        }
        ctx.fillStyle = "black";
        if (score == 10000) {
            ctx.fillText("Evaluation: " + score + " GGWP", 20, 20)
        } else if (score == -10000) {
            ctx.fillText("Evaluation: " + score + " RIP", 20, 20)
        } else {
            ctx.fillText("Evaluation: " + score, 20, 20)
        }
        pile.sort(compareRank)
        dragCard(mousedown, mouse);
        drawCPU(p2_draw);
        drawPlayer(p1_draw, p1_plays);
        for (var i = 0; i < pile.length; i++) {
            pile[i].draw();
        }
        blueBorder(p1_plays);
    }
}

update();